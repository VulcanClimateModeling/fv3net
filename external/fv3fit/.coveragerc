[run]
omit = tests/* setup.py fv3fit/sklearn/__main__.py fv3fit/keras/__main__.py
