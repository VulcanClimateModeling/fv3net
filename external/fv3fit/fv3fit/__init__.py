from ._shared.predictor import Predictor, Estimator
from ._shared.io import dump, load
from . import keras
from . import sklearn