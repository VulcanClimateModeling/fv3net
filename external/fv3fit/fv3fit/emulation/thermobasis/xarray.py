import dataclasses
import logging
from typing import MutableMapping, Hashable, Union
import xarray as xr
import tensorflow as tf
from fv3fit.emulation.thermobasis.batch import to_tensors, to_dict_no_static_vars
from fv3fit.emulation.thermobasis.emulator import Trainer, Config

State = MutableMapping[Hashable, xr.DataArray]


@dataclasses.dataclass
class XarrayEmulator:
    """Wrap an OnlineEmulator to allow for xarrray inputs and outputs
    """

    emulator: Trainer

    @property
    def input_variables(self):
        return self.emulator.input_variables

    @property
    def output_variables(self):
        return self.emulator.output_variables

    def dump(self, path: str):
        return self.emulator.dump(path)

    @classmethod
    def load(cls, path: str):
        return cls(Trainer.load(path))

    def predict(self, state: State) -> State:
        in_ = stack(state, self.input_variables)
        in_tensors = to_tensors(in_)
        x = self.emulator.batch_to_specific_humidity_basis(in_tensors)
        out = self.emulator.model(x)

        tensors = to_dict_no_static_vars(out)

        dims = ["sample", "z"]
        attrs = {"units": "no one cares"}

        return dict(
            xr.Dataset(
                {key: (dims, val, attrs) for key, val in tensors.items()},
                coords=in_.coords,
            ).unstack("sample")
        )

    def partial_fit(self, statein: State, stateout: State):

        in_tensors = _xarray_to_tensor(statein, self.input_variables)
        out_tensors = _xarray_to_tensor(stateout, self.output_variables)
        d = tf.data.Dataset.from_tensor_slices((in_tensors, out_tensors)).shuffle(
            1_000_000
        )
        self.emulator.batch_fit(d)


def _xarray_to_tensor(state, keys):
    in_ = stack(state, keys)
    return to_tensors(in_)


def stack(state: State, keys) -> xr.Dataset:
    ds = xr.Dataset({key: state[key] for key in keys})
    sample_dims = ["y", "x"]
    return ds.stack(sample=sample_dims).transpose("sample", ...)


def get_xarray_emulator(config: Union[Config, str]) -> XarrayEmulator:
    if isinstance(config, str):
        logging.info(f"Loading emulator from checkpoint {config}")
        return XarrayEmulator(Trainer.load(config))
    else:
        return XarrayEmulator(Trainer(config))
