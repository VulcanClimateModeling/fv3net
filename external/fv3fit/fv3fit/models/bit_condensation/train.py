from fv3fit._shared import io
import tensorflow as tf
from typing import Optional
import numpy as np
from dataclasses import dataclass

from fv3fit._shared.training_config import Hyperparameters, register_training_function
from fv3fit._shared.predictor import Reloadable
from fv3fit.models.bit_condensation import integer_encoding


out_v = "specific_humidity_after_gscond"


# IEEE encoding
def encode_ieee(x):
    buf = x.astype(np.float16).tobytes()
    byte_array = np.frombuffer(buf, dtype=np.uint8)
    n = np.unpackbits(byte_array, axis=-1)
    shape = x.shape + (-1,)
    return n.reshape(shape)

    # IEEE encoding


def decode_ieee(y):
    uint8 = np.packbits(y, -1)
    buf = uint8.tobytes()
    return np.frombuffer(buf, np.float16)


# tests
def test_ieee():
    x = np.array([9.83839], dtype=np.float16)
    x_back = decode_ieee(encode_ieee(x))
    np.testing.assert_array_equal(x_back, x)


assert encode_ieee(np.array([1.0])).shape == (1, 16)


@dataclass
class Config(Hyperparameters):
    depth: int = 2
    width: int = 256
    batch_size: int = 2 ** 14
    epochs: int = 1
    overfit: bool = False

    @property
    def variables(self):
        return set(get_inputs()) | {out_v}


def get_inputs():
    return [
        "air_pressure",
        "surface_air_pressure",
        "surface_air_pressure_after_last_gscond",
        "specific_humidity_input",
        "specific_humidity_after_last_gscond",
        "air_temperature_input",
        "air_temperature_after_last_gscond",
        "pressure_thickness_of_atmospheric_layer",
        "cloud_water_mixing_ratio_input",
    ]


def stack(d):
    p = d["air_pressure"]
    ps = d["surface_air_pressure"][..., None]
    ps_last = d["surface_air_pressure_after_last_gscond"][..., None]
    p_last = p / ps * ps_last
    cloud = tf.math.maximum(1e-20, d["cloud_water_mixing_ratio_input"])
    return tf.stack(
        [
            d["specific_humidity_input"],
            tf.math.log(d["specific_humidity_input"]),
            d["specific_humidity_after_last_gscond"],
            d["air_temperature_input"],
            d["air_temperature_after_last_gscond"],
            p,
            p_last,
            d["pressure_thickness_of_atmospheric_layer"],
            cloud,
            tf.math.log(cloud),
        ],
        -1,
    )


def get_outputs(d):
    return d["specific_humidity_after_gscond"] - d["specific_humidity_input"]


def preprocess(d):
    inputs = stack(d)
    output = get_outputs(d)
    return inputs, output


def encode(x):
    in_bits = tf.py_function(lambda x: encode_ieee(x.numpy()), [x], tf.bool)
    return tf.ensure_shape(in_bits, list(x.shape) + [16])


def decode(y):
    out = tf.py_function(lambda y: decode_ieee(y.numpy()), [y], tf.float16)
    return tf.ensure_shape(out, y.shape[:-1])


def encode_xy(x, y):
    return encode(x), encode(y)


@io.register("bit-condensation")
class KerasModel(Reloadable):
    def __init__(self, model):
        self.model = model

    def dump(self, path: str) -> None:
        """Serialize to a directory."""
        self.model.save(path + "/model.tf")

    @staticmethod
    def load(path: str) -> "KerasModel":
        model = tf.keras.models.load_model(path + "/model.tf")
        return KerasModel(model)

    def predict_numpy(self, d):
        stacked = stack(d).numpy()
        encoded = encode_ieee(stacked)
        encoded_flat = tf.reshape(encoded, [-1, 8, 16])
        y_bits = self.model.predict(encoded_flat) > 0
        y = decode_ieee(y_bits)
        return y.reshape([-1, 79])


class EndPoint(tf.keras.layers.Layer):
    def __init__(self, coder):
        super().__init__()
        self.loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=True)
        self.coder = coder
        self.dense = tf.keras.layers.Dense(coder.size)

    def call(self, hidden, targets=None):
        logits = tf.squeeze(self.dense(hidden))
        if targets is not None:
            bits = self.coder.encode(targets)
            y_hat = self.coder.decode(logits)

            y_hat = tf.cast(y_hat, tf.float32)
            targets = tf.cast(targets, tf.float32)

            target_decoded = self.coder.decode(self.coder.encode(targets))

            mse = tf.keras.metrics.mean_squared_error(targets, y_hat)
            mse_null = tf.keras.metrics.mean_squared_error(
                targets, tf.zeros_like(targets)
            )
            mse_coding = tf.keras.metrics.mean_squared_error(targets, target_decoded)
            r2 = 1 - mse / mse_null
            self.add_metric(mse, aggregation="mean", name="mse")
            self.add_metric(mse_coding, aggregation="mean", name="mse_coding")
            self.add_metric(mse_null, aggregation="mean", name="mse_null")
            self.add_metric(r2, aggregation="mean", name="r2")
            loss = self.coder.loss(bits, logits)
            self.add_loss(loss)
        return tf.nn.sigmoid(logits)


@register_training_function("bit_condensation", Config)
def train(
    hyperparameters: Config,
    train_batches: tf.data.Dataset,
    validation_batches: Optional[tf.data.Dataset],
) -> KerasModel:
    b = train_batches.map(preprocess)

    grid_points = b.unbatch().unbatch()

    if hyperparameters.overfit:
        grid_points = grid_points.take(10)
    else:
        grid_points = grid_points.shuffle(10_000)

    grid_points = grid_points.batch(hyperparameters.batch_size).map(
        lambda x, y: {"x": x, "y": y}
    )
    input_norm = tf.keras.layers.Normalization(axis=1)
    input_norm.adapt(grid_points.map(lambda x: x["x"]).take(10))

    x = tf.keras.Input(10, name="x")
    y = tf.keras.Input([], name="y")
    x_norm = input_norm(x)

    state = x_norm
    for i in range(hyperparameters.depth):
        dense = tf.keras.layers.Dense(hyperparameters.width, activation="relu")
        state = dense(state)

    # endpoint pattern
    # https://keras.io/examples/keras_recipes/endpoint_layer_pattern/
    dt = tf.int16

    coder = integer_encoding.IntEncoder(1e-5, dt=dt)
    coder = integer_encoding.OneHot(1e-3)
    coder = integer_encoding.Float()
    coder = integer_encoding.Mixed()
    endpoint = EndPoint(coder)
    preds = endpoint(state, y)
    model = tf.keras.Model([x, y], preds)
    model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3))
    model.fit(grid_points, epochs=hyperparameters.epochs)
    return KerasModel(model)
