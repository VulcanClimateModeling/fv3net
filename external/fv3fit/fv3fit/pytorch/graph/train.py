import tensorflow as tf
import numpy as np
import dataclasses
from fv3fit._shared.training_config import Hyperparameters
from toolz.functoolz import curry
from fv3fit.pytorch.predict import PytorchModel
from fv3fit.pytorch.graph.network import GraphNetwork, GraphNetworkConfig
from fv3fit.pytorch.loss import LossConfig
from fv3fit.pytorch.optimizer import OptimizerConfig
from fv3fit.pytorch.training_loop import TrainingLoopConfig
from fv3fit._shared.scaler import TensorStandardScaler
from ..system import DEVICE

from fv3fit._shared import register_training_function
from typing import (
    List,
    Optional,
    Sequence,
    Set,
    Mapping,
)
from fv3fit.tfdataset import select_keys, ensure_nd, apply_to_mapping


@dataclasses.dataclass
class GraphHyperparameters(Hyperparameters):
    """
    Args:
        state_variables: names of variables to evolve forward in time
        optimizer_config: selection of algorithm to be used in gradient descent
        graph_network: configuration of graph network
        training_loop: configuration of training loop
        loss: configuration of loss functions, will be applied separately to
            each output variable
    """

    state_variables: List[str]
    normalization_fit_samples: int = 50_000
    optimizer_config: OptimizerConfig = dataclasses.field(
        default_factory=lambda: OptimizerConfig("AdamW")
    )
    graph_network: GraphNetworkConfig = dataclasses.field(
        default_factory=lambda: GraphNetworkConfig()
    )
    training_loop: TrainingLoopConfig = dataclasses.field(
        default_factory=lambda: TrainingLoopConfig()
    )
    loss: LossConfig = LossConfig(loss_type="mse")

    @property
    def variables(self) -> Set[str]:
        return set(self.state_variables)


def get_normalizer(sample: Mapping[str, np.ndarray]):
    scalers = {}
    for name, array in sample.items():
        s = TensorStandardScaler(n_sample_dims=5)
        s.fit(array)
        scalers[name] = s

    def scale(sample: Mapping[str, np.ndarray]):
        output = {**sample}
        for name, array in sample.items():
            output[name] = scalers[name].normalize(array)
        return output

    return scale


# TODO: Still have to handle forcing


@register_training_function("graph", GraphHyperparameters)
def train_graph_model(
    hyperparameters: GraphHyperparameters,
    train_batches: tf.data.Dataset,
    validation_batches: Optional[tf.data.Dataset],
):
    """
    Train a graph network.

    Args:
        hyperparameters: configuration for training
        train_batches: training data, as a dataset of Mapping[str, tf.Tensor]
            where each tensor has dimensions [batch, time, tile, x, y(, z)]
        validation_batches: validation data, as a dataset of Mapping[str, tf.Tensor]
            where each tensor has dimensions [batch, time, tile, x, y(, z)]
    """
    train_batches = train_batches.map(apply_to_mapping(ensure_nd(6)))
    sample = next(
        iter(train_batches.unbatch().batch(hyperparameters.normalization_fit_samples))
    )

    normalizer = get_normalizer(sample)

    get_state = curry(get_Xy_dataset)(
        state_variables=hyperparameters.state_variables,
        n_dims=6,  # [batch, time, tile, x, y, z]
        normalizer=normalizer,
    )

    if validation_batches is not None:
        val_state = get_state(data=validation_batches)
    else:
        val_state = None

    train_state = get_state(data=train_batches)

    train_model = build_model(
        hyperparameters.graph_network, n_state=next(iter(train_state)).shape[-1]
    )
    optimizer = hyperparameters.optimizer_config

    hyperparameters.training_loop.fit_loop(
        hyperparameters.loss,
        train_model=train_model,
        train_data=train_state,
        validation_data=val_state,
        optimizer=optimizer.instance(train_model.parameters()),
    )

    predictor = PytorchModel(
        input_variables=hyperparameters.state_variables,
        output_variables=hyperparameters.state_variables,
        model=train_model,
        normalizers=normalizer,
        unstacked_dims=("time", "tile", "x", "y", "z"),
    )
    return predictor


def build_model(graph_network, n_state: int):
    """
    Args:
        graph_network: configuration of the graph network
        n_state: number of state variables
    """
    train_model = GraphNetwork(
        graph_network, n_features_in=n_state, n_features_out=n_state
    ).to(DEVICE)
    return train_model


def get_Xy_dataset(
    state_variables: Sequence[str], n_dims: int, normalizer, data: tf.data.Dataset,
):
    """
    Given a tf.data.Dataset with mappings from variable name to samples of shape
    [batch, time, tile, x, y(, z)]
    return a tf.data.Dataset whose entries are tensors of the requested
    state variables concatenated along the feature dimension, of shape
    [batch, time, tile, x, y, feature].
    """
    ensure_dims = apply_to_mapping(ensure_nd(n_dims))

    def map_fn(data):
        data = normalizer(data)
        data = ensure_dims(data)
        data = select_keys(state_variables, data)
        data = tf.concat(data, axis=-1)
        return data

    return data.map(map_fn)
