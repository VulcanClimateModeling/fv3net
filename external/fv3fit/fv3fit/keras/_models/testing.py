from typing import Iterable, Sequence, Optional
import logging
import os
import xarray as xr
import numpy as np
from .models import Model, _XyArraySequence
from ._filesystem import get_dir, put_dir
from ..._shared import ArrayPacker

logger = logging.getLogger(__file__)


class DummyModel(Model):
    """
    A dummy keras model for testing, whose `fit` method learns only the input and
    output variable array dimensions in an xarray dataset and ignores their contents,
    and which simply returns zeros for all output variable features
    """

    # these should only be used in the dump/load routines for this class
    _X_PACKER_FILENAME = "X_packer.json"
    _Y_PACKER_FILENAME = "y_packer.json"

    def __init__(
        self,
        sample_dim_name: str,
        input_variables: Iterable[str],
        output_variables: Iterable[str],
    ):
        """Initialize the DummyModel
        Args:
            sample_dim_name: name of the sample dimension in datasets used as
                inputs and outputs.
            input_variables: names of input variables
            output_variables: names of output variables
        """
        super().__init__(sample_dim_name, input_variables, output_variables)
        self.X_packer = ArrayPacker(
            sample_dim_name=sample_dim_name, pack_names=input_variables
        )
        self.y_packer = ArrayPacker(
            sample_dim_name=sample_dim_name, pack_names=output_variables
        )

    def fit(
        self,
        batches: Sequence[xr.Dataset],
        batch_size: Optional[int] = None,
        epochs: Optional[int] = None,
    ) -> None:
        # this is all we need to do to learn n output feature
        _, _ = _XyArraySequence(self.X_packer, self.y_packer, batches)[0]

    def predict(self, X: xr.Dataset) -> xr.Dataset:
        if not self.y_packer._n_features:
            raise RuntimeError("must call fit() for dummy model to be available")
        feature_index = X[self.sample_dim_name]
        ds_pred = self.y_packer.to_dataset(
            self.predict_array(self.X_packer.to_array(X))
        )
        return ds_pred.assign_coords({self.sample_dim_name: feature_index})

    def predict_array(self, X: np.ndarray) -> np.ndarray:
        return np.zeros((X.shape[0], self.y_packer._total_features))

    def dump(self, path: str) -> None:
        with put_dir(path) as path:
            with open(os.path.join(path, self._X_PACKER_FILENAME), "w") as f:
                self.X_packer.dump(f)
            with open(os.path.join(path, self._Y_PACKER_FILENAME), "w") as f:
                self.y_packer.dump(f)

    @classmethod
    def load(cls, path: str) -> Model:
        with get_dir(path) as path:
            with open(os.path.join(path, cls._X_PACKER_FILENAME), "r") as f:
                X_packer = ArrayPacker.load(f)
            with open(os.path.join(path, cls._Y_PACKER_FILENAME), "r") as f:
                y_packer = ArrayPacker.load(f)
            obj = cls(
                X_packer.sample_dim_name, X_packer.pack_names, y_packer.pack_names
            )
            obj.X_packer = X_packer
            obj.y_packer = y_packer
            return obj
