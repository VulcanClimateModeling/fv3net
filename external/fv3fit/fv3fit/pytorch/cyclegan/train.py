from fv3fit._shared.hyperparameters import Hyperparameters
import dataclasses
import tensorflow as tf
from fv3fit.pytorch.predict import PytorchPredictor
from fv3fit.pytorch.loss import LossConfig
from fv3fit.pytorch.optimizer import OptimizerConfig
import tensorflow_datasets as tfds
from fv3fit.tfdataset import sequence_size
import torch
import numpy as np
from ..system import DEVICE

from fv3fit._shared import register_training_function
from typing import (
    List,
    Optional,
    Tuple,
)
from fv3fit.tfdataset import ensure_nd, apply_to_mapping
from .network import Generator
from fv3fit.pytorch.graph.train import (
    get_scalers,
    get_mapping_scale_func,
    get_Xy_dataset,
)
from toolz import curry
import logging

logger = logging.getLogger(__name__)


@dataclasses.dataclass
class GeneratorConfig:
    n_convolutions: int = 3
    n_resnet: int = 3
    max_filters: int = 256


@dataclasses.dataclass
class AutoencoderHyperparameters(Hyperparameters):

    state_variables: List[str]
    normalization_fit_samples: int = 50_000
    optimizer_config: OptimizerConfig = dataclasses.field(
        default_factory=lambda: OptimizerConfig("AdamW")
    )
    generator: GeneratorConfig = dataclasses.field(
        default_factory=lambda: GeneratorConfig()
    )
    training_loop: "TrainingLoopConfig" = dataclasses.field(
        default_factory=lambda: TrainingLoopConfig()
    )
    loss: LossConfig = LossConfig(loss_type="mse")

    @property
    def variables(self):
        return tuple(self.state_variables)


@dataclasses.dataclass
class TrainingLoopConfig:
    """
    Attributes:
        epochs: number of times to run through the batches when training
        shuffle_buffer_size: size of buffer to use when shuffling samples
        save_path: name of the file to save the best weights
        do_multistep: if True, use multistep loss calculation
        multistep: number of steps in multistep loss calculation
        validation_batch_size: if given, process validation data in batches
            of this size, otherwise process it all at once
    """

    n_epoch: int = 20
    shuffle_buffer_size: int = 10
    samples_per_batch: int = 1
    save_path: str = "weight.pt"
    validation_batch_size: Optional[int] = None

    def fit_loop(
        self,
        train_model: torch.nn.Module,
        train_data: tf.data.Dataset,
        validation_data: tf.data.Dataset,
        optimizer: torch.optim.Optimizer,
        loss_config: LossConfig,
    ) -> None:
        """
        Args:
            train_model: pytorch model to train
            train_data: training dataset containing samples to be passed to the model,
                samples should be tuples with two tensors of shape
                [sample, time, tile, x, y, z]
            validation_data: validation dataset containing samples to be passed
                to the model, samples should be tuples with two tensors
                of shape [sample, time, tile, x, y, z]
            optimizer: type of optimizer for the model
            loss_config: configuration of loss function
        """
        train_data = (
            flatten_dims(train_data)
            .shuffle(buffer_size=self.shuffle_buffer_size)
            .batch(self.samples_per_batch)
        )
        train_data = tfds.as_numpy(train_data)
        if validation_data is not None:
            if self.validation_batch_size is None:
                validation_batch_size = sequence_size(validation_data)
            else:
                validation_batch_size = self.validation_batch_size
            validation_data = flatten_dims(validation_data).batch(validation_batch_size)
            validation_data = tfds.as_numpy(validation_data)
            min_val_loss = np.inf
            best_weights = None
        for i in range(1, self.n_epoch + 1):  # loop over the dataset multiple times
            logger.info("starting epoch %d", i)
            train_model = train_model.train()
            train_losses = []
            for batch_state in train_data:
                batch_input = torch.as_tensor(batch_state[0]).float().to(DEVICE)
                batch_output = torch.as_tensor(batch_state[1]).float().to(DEVICE)
                optimizer.zero_grad()
                loss: torch.Tensor = loss_config.loss(
                    train_model(batch_input), batch_output
                )
                loss.backward()
                train_losses.append(loss)
                optimizer.step()
            train_loss = torch.mean(torch.stack(train_losses))
            logger.info("train loss: %f", train_loss)
            if validation_data is not None:
                val_model = train_model.eval()
                val_losses = []
                for batch_state in validation_data:
                    batch_input = torch.as_tensor(batch_state[0]).float().to(DEVICE)
                    batch_output = torch.as_tensor(batch_state[1]).float().to(DEVICE)
                    with torch.no_grad():
                        val_losses.append(
                            loss_config.loss(val_model(batch_input), batch_output)
                        )
                val_loss = torch.mean(torch.stack(val_losses))
                logger.info("val_loss %f", val_loss)
                if val_loss < min_val_loss:
                    min_val_loss = val_loss
                    best_weights = train_model.state_dict()
        if validation_data is not None:
            train_model.load_state_dict(best_weights)


@curry
def define_noisy_input(data: tf.Tensor, stdev=0.1) -> Tuple[tf.Tensor, tf.Tensor]:
    """
    Given data, return a tuple with a noisy version of the data and the original data.
    """
    noisy = data + tf.random.normal(shape=tf.shape(data), stddev=stdev)
    return (noisy, data)


def flatten_dims(dataset: tf.data.Dataset) -> tf.data.Dataset:
    """Transform [batch, time, tile, x, y, z] to [sample, x, y, z]"""
    return dataset.unbatch().unbatch().unbatch()


@register_training_function("autoencoder", AutoencoderHyperparameters)
def train_autoencoder(
    hyperparameters: AutoencoderHyperparameters,
    train_batches: tf.data.Dataset,
    validation_batches: Optional[tf.data.Dataset],
) -> PytorchPredictor:
    """
    Train a denoising autoencoder for cubed sphere data.

    Args:
        hyperparameters: configuration for training
        train_batches: training data, as a dataset of Mapping[str, tf.Tensor]
            where each tensor has dimensions [sample, time, tile, x, y(, z)]
        validation_batches: validation data, as a dataset of Mapping[str, tf.Tensor]
            where each tensor has dimensions [sample, time, tile, x, y(, z)]
    """
    train_batches = train_batches.map(apply_to_mapping(ensure_nd(6)))
    sample_batch = next(
        iter(train_batches.unbatch().batch(hyperparameters.normalization_fit_samples))
    )

    scalers = get_scalers(sample_batch)
    mapping_scale_func = get_mapping_scale_func(scalers)

    get_state = curry(get_Xy_dataset)(
        state_variables=hyperparameters.state_variables,
        n_dims=6,  # [batch, time, tile, x, y, z]
        mapping_scale_func=mapping_scale_func,
    )

    if validation_batches is not None:
        val_state = get_state(data=validation_batches)
    else:
        val_state = None

    train_state = get_state(data=train_batches)

    train_model = build_model(
        hyperparameters.generator, n_state=next(iter(train_state)).shape[-1]
    )
    print(train_model)
    optimizer = hyperparameters.optimizer_config

    train_state = train_state.map(define_noisy_input(stdev=0.5))
    if validation_batches is not None:
        val_state = val_state.map(define_noisy_input(stdev=0.5))

    hyperparameters.training_loop.fit_loop(
        train_model=train_model,
        train_data=train_state,
        validation_data=val_state,
        optimizer=optimizer.instance(train_model.parameters()),
        loss_config=hyperparameters.loss,
    )

    predictor = PytorchPredictor(
        input_variables=hyperparameters.state_variables,
        output_variables=hyperparameters.state_variables,
        model=train_model,
        scalers=scalers,
    )
    return predictor


def build_model(config: GeneratorConfig, n_state: int) -> Generator:
    return Generator(
        channels=n_state,
        n_convolutions=config.n_convolutions,
        n_resnet=config.n_resnet,
        max_filters=config.max_filters,
    )
