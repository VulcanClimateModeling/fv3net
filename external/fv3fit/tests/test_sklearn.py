from sklearn.ensemble import RandomForestRegressor
import numpy as np
import joblib
import fv3fit.train
import xarray as xr
import vcm.testing


def test_numpy_version(regtest):
    print(np.version.full_version, file=regtest)


def test_random_state_uniform_reproducibility():
    seed = 10
    fv3fit.set_random_seed(seed)
    random = np.random.RandomState(0)
    n_samples = 500
    X0 = random.uniform(size=(n_samples, 5))

    fv3fit.set_random_seed(seed)
    random = np.random.RandomState(0)
    n_samples = 500
    X1 = random.uniform(size=(n_samples, 5))
    assert np.array_equal(X0, X1)


def test_random_state_second_output_reproducibility(regtest):
    random = np.random.RandomState(0)
    n_samples = 500
    _ = random.uniform(size=(n_samples, 5))
    X = random.uniform(size=(n_samples, 5))
    print(joblib.hash(X), file=regtest)


def test_data_generation_reproducibility(regtest):
    """
    This sequence of operations occurs in another test, and at one point
    seemed not to be reproducible between systems.
    """
    fv3fit.set_random_seed(1)
    random = np.random.RandomState(0)
    n_sample, n_feature = int(5e3), 2

    def sample_func():
        return xr.DataArray(
            random.uniform(size=(n_sample, n_feature)), dims=["sample", "feature_dim"]
        )

    _ = sample_func()
    data_array = sample_func()
    test_dataset = xr.Dataset(data_vars={"var_in": data_array, "var_out": data_array})
    for result in vcm.testing.checksum_dataarray_mapping(test_dataset):
        print(result, file=regtest)


def set_random_state_train_and_predict(seed):
    fv3fit.set_random_seed(seed)
    regressor = RandomForestRegressor(random_state=0, n_jobs=None)
    random = np.random.RandomState(0)
    n_samples = 500
    X = random.uniform(size=(n_samples, 5))
    y = random.uniform(size=(n_samples, 2))
    with joblib.parallel_backend("loky", n_jobs=8):
        regressor.fit(X, y)
    X_test = random.uniform(size=(n_samples, 5))
    y_test = regressor.predict(X_test)
    return y_test


def test_random_forest_reproducibility():
    seed = 10
    prediction_0 = set_random_state_train_and_predict(seed)
    prediction_1 = set_random_state_train_and_predict(seed)
    assert np.array_equal(prediction_0, prediction_1)


def test_random_forest_n_jobs_can_exceed_n_estimators():
    regressor = RandomForestRegressor(random_state=0, n_estimators=1, n_jobs=None)
    random = np.random.RandomState(0)
    n_samples = 100
    X = random.uniform(size=(n_samples, 5))
    y = random.uniform(size=(n_samples, 2))
    with joblib.parallel_backend("loky", n_jobs=8):
        regressor.fit(X, y)
    X_test = random.uniform(size=(n_samples, 5))
    regressor.predict(X_test)
