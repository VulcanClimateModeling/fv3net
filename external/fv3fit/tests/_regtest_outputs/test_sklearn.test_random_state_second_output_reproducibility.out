9c5f6dd058d6c394c9404a9b0261e91e
